libfwupdplugin/fu-crc.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+//
+// SPDX-License-Identifier: LGPL-2.1+
+
+use std::mem;
+use std::slice;
+
+fn crc8_full(buf: &[u8], crc_init: u8, polynomial: u8) -> u8 {
+    let mut crc = crc_init as u32;
+    for tmp in buf.iter() {
+        crc ^= (*tmp as u32) << 8;
+        for _ in 0..8 {
+            if crc & 0x8000 > 0 {
+                crc ^= ((polynomial as u32 | 0x100) << 7) as u32;
+            }
+            crc <<= 1;
+        }
+    }
+    !(crc >> 8) as u8
+}
+
+fn crc16_full(buf: &[u8], crc_init: u16, polynomial: u16) -> u16 {
+    let mut crc = crc_init;
+    for tmp in buf.iter() {
+        crc = crc ^ *tmp as u16;
+        for _ in 0..8 {
+            if crc & 0x1 > 0 {
+                crc = (crc >> 1) ^ polynomial;
+            } else {
+                crc >>= 1;
+            }
+        }
+    }
+    !crc
+}
+
+fn crc32_full(buf: &[u8], crc_init: u32, polynomial: u32) -> u32 {
+    let mut crc = crc_init;
+    for tmp in buf.iter() {
+        crc = crc ^ *tmp as u32;
+        for _ in 0..8 {
+            let mask = {
+                if crc & 0b1 > 0 {
+                    0xFFFFFFFF
+                } else {
+                    0x0
+                }
+            };
+            crc = (crc >> 1) ^ (polynomial & mask);
+        }
+    }
+    !crc
+}
+
+#[no_mangle]
+pub extern "C" fn fu_crc8_full(buf: *mut u8, bufsz: usize, crc_init: u8, polynomial: u8) -> u8 {
+    let slice = unsafe { slice::from_raw_parts(buf, bufsz) };
+    let ret = crc8_full(&slice, crc_init, polynomial);
+    mem::forget(slice);
+    ret
+}
+
+#[no_mangle]
+pub extern "C" fn fu_crc16_full(buf: *mut u8, bufsz: usize, crc_init: u16, polynomial: u16) -> u16 {
+    let slice = unsafe { slice::from_raw_parts(buf, bufsz) };
+    let ret = crc16_full(&slice, crc_init, polynomial);
+    mem::forget(slice);
+    ret
+}
+
+#[no_mangle]
+pub extern "C" fn fu_crc32_full(buf: *mut u8, bufsz: usize, crc_init: u32, polynomial: u32) -> u32 {
+    let slice = unsafe { slice::from_raw_parts(buf, bufsz) };
+    let ret = crc32_full(&slice, crc_init, polynomial);
+    mem::forget(slice);
+    ret
+}
+
+#[test]
+fn fu_crc32() {
```

Comment:
this should rename, it's testing all 3 of the CRC functions in one run

Reply:
Fixed.